File: ios/Classes/common/Serialization.swift
import HyperTrack

private let keyType = "type"
private let keyValue = "value"

private let typeLocation = "location"
private let typeSuccess = "success"
private let typeFailure = "failure"
private let typeHyperTrackError = "hyperTrackError"
private let typeLocationErrorNotRunning = "notRunning"
private let typeLocationErrorStarting = "starting"
private let typeLocationErrorErrors = "errors"
private let typeIsTracking = "isTracking"
private let typeIsAvailable = "isAvailable"

private let keyLatitude = "latitude"
private let keyLongitude = "longitude"

let keyGeotagData = "data"
let keyPublishableKey = "publishableKey"

func getHyperTrackError(_ error: HyperTrack.UnrestorableError) -> HyperTrackError {
    switch(error) {
    case .invalidPublishableKey:
        return HyperTrackError.invalidPublishableKey
    case .motionActivityPermissionsDenied:
        return HyperTrackError.locationPermissionsDenied
    }
}

func getHyperTrackError(_ error: HyperTrack.RestorableError) -> HyperTrackError {
    switch(error) {
    case .locationPermissionsNotDetermined:
        return HyperTrackError.locationPermissionsNotDetermined
    case .motionActivityPermissionsNotDetermined:
        return HyperTrackError.motionActivityPermissionsNotDetermined
    case .locationPermissionsCantBeAskedInBackground:
        return HyperTrackError.locationPermissionsNotDetermined
    case .motionActivityPermissionsCantBeAskedInBackground:
        return HyperTrackError.motionActivityPermissionsNotDetermined
    case .locationPermissionsRestricted:
        return HyperTrackError.locationPermissionsRestricted
    case .motionActivityPermissionsRestricted:
        return HyperTrackError.motionActivityPermissionsRestricted
    case .locationPermissionsDenied:
        return HyperTrackError.locationPermissionsDenied
    case .locationPermissionsInsufficientForBackground:
        return HyperTrackError.locationPermissionsInsufficientForBackground
    case .locationServicesDisabled:
        return HyperTrackError.locationServicesDisabled
    case .motionActivityServicesDisabled:
        return HyperTrackError.motionActivityServicesDisabled
    case .networkConnectionUnavailable:
        return HyperTrackError.networkConnectionUnavailable
    case .trialEnded:
        return HyperTrackError.blockedFromRunning
    case .paymentDefault:
        return HyperTrackError.blockedFromRunning
    }
}

func serializeFatalError(fatalError: HyperTrack.FatalError) -> String {
    switch fatalError {
    case .developmentError(.missingLocationUpdatesBackgroundModeCapability):
        return "missingLocationUpdatesBackgroundModeCapability"
    case .productionError(.locationServicesUnavalible):
        return "locationServicesUnavalible"
    case .developmentError(.runningOnSimulatorUnsupported):
        return "runningOnSimulatorUnsupported"
    case .productionError(.motionActivityServicesUnavalible):
        return "motionActivityServicesUnavalible"
    case .productionError(.motionActivityPermissionsDenied):
        return "motionActivityPermissionsDenied"
    }
}

func serializeLocationResult(_ result: Result<HyperTrack.Location, HyperTrack.LocationError>) -> Dictionary<String, Any>  {
    switch (result) {
    case .success(let success):
        return [
            keyType: typeSuccess,
            keyValue: [
                keyType: typeLocation,
                keyValue: [
                    keyLatitude: success.latitude,
                    keyLongitude: success.longitude
                ]
            ]
        ]
    case .failure(let failure):
        var locationError: Dictionary<String, Any>
        switch(failure) {
        case .locationPermissionsNotDetermined:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsNotDetermined)
        case .locationPermissionsCantBeAskedInBackground:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsNotDetermined)
        case .locationPermissionsInsufficientForBackground:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsInsufficientForBackground)
        case .locationPermissionsRestricted:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsRestricted)
        case .locationPermissionsReducedAccuracy:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsReducedAccuracy)
        case .locationPermissionsDenied:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationPermissionsDenied)
        case .locationServicesDisabled:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationServicesDisabled)
        case .motionActivityPermissionsNotDetermined:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.motionActivityPermissionsNotDetermined)
        case .motionActivityPermissionsCantBeAskedInBackground:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.motionActivityPermissionsNotDetermined)
        case .motionActivityPermissionsDenied:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.motionActivityPermissionsDenied)
        case .motionActivityServicesDisabled:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.motionActivityServicesDisabled)
            
        case .gpsSignalLost:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.gpsSignalLost)
        case .locationMocked:
            locationError = serializeHyperTrackErrorAsLocationError(HyperTrackError.locationMocked)
        case .starting:
            locationError = [
                keyType: typeLocationErrorStarting
            ]
        case .notRunning:
            locationError = [
                keyType: typeLocationErrorNotRunning
            ]
        }
        
        return [
            keyType: typeFailure,
            keyValue: locationError
        ]
    }
}

func serializeHyperTrackErrorAsLocationError(_ error: HyperTrackError) -> Dictionary<String, Any> {
    return [
        keyType: typeLocationErrorErrors,
        keyValue: [
            [
                keyType: typeHyperTrackError,
                keyValue: error.rawValue
            ]
        ]
    ]
}

func serializeHyperTrackError(_ error: HyperTrackError) -> Dictionary<String, Any> {
    return [
        keyType: typeHyperTrackError,
        keyValue: error.rawValue
    ]
}

func serializeIsTracking(_ isTracking: Bool) -> Dictionary<String, Any> {
    return [
        keyType: typeIsTracking,
        keyValue: isTracking
    ]
}

func serializeIsAvailable(_ isAvailable: HyperTrack.Availability) -> Dictionary<String, Any> {
    return [
        keyType: typeIsAvailable,
        keyValue: isAvailable == .available
    ]
}

func deserializeAvailability(_ data: Dictionary<String, Any>) -> Result<Bool, FailureResult> {
    if(data[keyType] as? String != typeIsAvailable) {
        return .failure(.fatalError(getParseError(data, key: keyType)))
    }
    guard let value = data[keyValue] as? Bool else {
        return .failure(.fatalError(getParseError(data, key: keyValue)))
    }
    return .success(value)
}

func deserializeGeotagData(
    _ data: Dictionary<String, Any>
) -> Result<Dictionary<String, Any>, FailureResult> {
    guard let data = data[keyGeotagData] as? Dictionary<String, Any> else {
        return .failure(.fatalError(getParseError(data, key: keyGeotagData)))
    }
    return .success(data)
}

func getParseError(_ data: Any, key: String) -> String {
    return "Invalid input for key \(key): \(data)"
}

File: android/src/main/kotlin/com/hypertrack/sdk/flutter/common/Serialization.kt
package com.hypertrack.sdk.flutter.common

import android.location.Location
import java.lang.IllegalArgumentException

/**
 * Platform-independent serialization code that converts HyperTrack data types
 * to Map<String, T> or List<T> where T is any JSON-compatible type
 */

fun serializeErrors(errors: Set<HyperTrackError>): List<Map<String, String>> {
  return errors.map {
    serializeHypertrackError(it)
  }
}

fun serializeSuccess(location: Location): Map<String, Any> {
  return mapOf(
    KEY_TYPE to TYPE_RESULT_SUCCESS,
    KEY_VALUE to serializeLocation(location)
  )
}

fun serializeFailure(locationError: LocationError): Map<String, Any> {
  return mapOf(
    KEY_TYPE to TYPE_RESULT_FAILURE,
    KEY_VALUE to serializeLocationError(locationError)
  )
}

fun serializeIsTracking(isTracking: Boolean): Map<String, Any> {
  return mapOf(
    KEY_TYPE to TYPE_IS_TRACKING,
    KEY_VALUE to isTracking
  )
}

fun serializeIsAvailable(isAvailable: Boolean): Map<String, Any> {
  return mapOf(
    KEY_TYPE to TYPE_AVAILABILITY,
    KEY_VALUE to isAvailable
  )
}

fun deserializeAvailability(isAvailable: Map<String, Any>): Boolean {
  if(isAvailable.getValue(KEY_TYPE) != TYPE_AVAILABILITY) {
    throw IllegalArgumentException(isAvailable.toString())
  }
  return isAvailable.getValue(KEY_VALUE) as Boolean
}

fun serializeLocation(location: Location): Map<String, Double> {
  return mapOf(
    KEY_LATITUDE to location.latitude,
    KEY_LONGITUDE to location.longitude
  )
}

fun serializeLocationError(locationError: LocationError): Map<String, Any> {
  return when(locationError) {
    NotRunning -> {
      mapOf(KEY_TYPE to TYPE_LOCATION_ERROR_NOT_RUNNING)
    }
    Starting -> {
      mapOf(KEY_TYPE to TYPE_LOCATION_ERROR_STARTING)
    }
    is Errors -> {
      mapOf(
        KEY_TYPE to TYPE_LOCATION_ERROR_ERRORS,
        KEY_VALUE to locationError.errors.map { serializeHypertrackError(it) }
      )
    }
  }
}

fun serializeHypertrackError(error: HyperTrackError): Map<String, String> {
  return mapOf(
    KEY_TYPE to TYPE_HYPERTRACK_ERROR,
    KEY_VALUE to error.name
  )
}

private const val KEY_TYPE = "type"
private const val KEY_VALUE = "value"

private const val TYPE_RESULT_SUCCESS = "success"
private const val TYPE_RESULT_FAILURE = "failure"

private const val TYPE_AVAILABILITY = "isAvailable"
private const val TYPE_HYPERTRACK_ERROR = "hyperTrackError"
private const val TYPE_IS_TRACKING = "isTracking"

private const val TYPE_LOCATION_ERROR_NOT_RUNNING = "notRunning"
private const val TYPE_LOCATION_ERROR_STARTING = "starting"
private const val TYPE_LOCATION_ERROR_ERRORS = "errors"

private const val KEY_LATITUDE = "latitude"
private const val KEY_LONGITUDE = "longitude"

const val KEY_GEOTAG_DATA = "data"
